src/java/org/apache/cassandra/index/sai/IndexContext.java

```diff
@@ -0,0 +1,435 @@
+import org.apache.cassandra.dht.AbstractBounds;
+import org.apache.cassandra.index.sai.analyzer.AbstractAnalyzer;
+import org.apache.cassandra.index.sai.memory.MemtableIndex;
+import org.apache.cassandra.index.sai.metrics.IndexMetrics;
+import org.apache.cassandra.index.sai.plan.Expression;
+import org.apache.cassandra.index.sai.utils.PrimaryKey;
+import org.apache.cassandra.index.sai.utils.PrimaryKeyFactory;
+import org.apache.cassandra.index.sai.utils.RangeIterator;
+import org.apache.cassandra.index.sai.utils.RangeUnionIterator;
+import org.apache.cassandra.index.sai.utils.TypeUtil;
+import org.apache.cassandra.schema.ColumnMetadata;
+import org.apache.cassandra.schema.IndexMetadata;
+import org.apache.cassandra.utils.Clock;
+import org.apache.cassandra.utils.FBUtilities;
+
+/**
+ * Manage metadata for each column index.
+ */
+public class IndexContext
+{
+    private static final Set<AbstractType<?>> EQ_ONLY_TYPES = ImmutableSet.of(UTF8Type.instance,
+                                                                              AsciiType.instance,
+                                                                              BooleanType.instance,
+                                                                              UUIDType.instance);
+
+    private final AbstractType<?> partitionKeyType;
+    private final ClusteringComparator clusteringComparator;
+
+    private final String keyspace;
+    private final String table;
+    private final ColumnMetadata columnMetadata;
+    private final IndexTarget.Type indexType;
+    private final AbstractType<?> validator;
+
+    // Config can be null if the column context is "fake" (i.e. created for a filtering expression).
+    private final IndexMetadata config;
+
+    private final ConcurrentMap<Memtable, MemtableIndex> liveMemtables = new ConcurrentHashMap<>();
+
+    private final IndexMetrics indexMetrics;
+
+    private final AbstractAnalyzer.AnalyzerFactory indexAnalyzerFactory;
+    private final AbstractAnalyzer.AnalyzerFactory queryAnalyzerFactory;
+    private final PrimaryKeyFactory primaryKeyFactory;
+
+    public IndexContext(String keyspace,
+                        String table,
+                        AbstractType<?> partitionKeyType,
+                        ClusteringComparator clusteringComparator,
+                        ColumnMetadata columnMetadata,
+                        IndexTarget.Type indexType,
+                        @Nullable IndexMetadata config)
+    {
+        this.keyspace = Objects.requireNonNull(keyspace);
+        this.table = Objects.requireNonNull(table);
+        this.partitionKeyType = Objects.requireNonNull(partitionKeyType);
+        this.clusteringComparator = Objects.requireNonNull(clusteringComparator);
+        this.columnMetadata = Objects.requireNonNull(columnMetadata);
+        this.indexType = Objects.requireNonNull(indexType);
+
+        this.config = config;
+        this.indexMetrics = new IndexMetrics(this);
```

Comment:
If this is a fake context, it seems like there are a few things we might be able to avoid. For example, creating an `IndexMetrics` and `liveMemtables` map seems unnecessary. Then looking at where those are used, say in `index()`, we could probably throw in an `assert this.config != null : "Attempted to index on a fake context!"` or something similar.

Reply:
This isn't a fake context. I removed all the fake context constructors in a previous set of changes.

Reply:
There is still a place where we create `IndexContext` w/ a `null` `config?`

Reply:
Right, I missed the point of your original comment as well. I've had another look at this and it bothers me that we have to have an `IndexContext` created like this. The sole usage of the `QueryController.getContext` is to get an analyzer from the context. It seems to make more sense for the `QueryController` to just return a `NoOpAnalyzer` if there is no index.